### ISSUE TITLE
Order of labelling functions affects the end model's performance

### ISSUE THREAD

**Issue by tienduccao** [HazyResearch/snorkel#948]
Hi, first of all thanks for creating this fantastic library.
I'm working on a NER problem using TagRNN as the end model.
I have a list of 10 labelling functions `LFs = lf_1, ..., lf_10`.
When I tried to reproduce the result of my model, I changed the order of `LFs` by mistake and obtain the very different results.

1.  For the original order of LFs, I obtained 0.741 F1 score on dev set of 0.706 on test set.
2. When I swapped the last 3 labelling functions (`lf_8, lf_9, lf_10 = original_lf_10, original_lf_8, original_lf_9`) I got 0.667 on dev and 0.622 on test.
3. When I placed the `original_lf_10` as the new `lf_1` and the original functions from `lf_1` to `lf_9` became the new `lf_2` to `lf_10`, I got 0.783 on dev and 0.676 on test.

So I think that the order of labelling functions creates different probabilistic labels for training set and it leads to different performances of the end model.
Do you have any comment on this issue?
Thanks.

**Comment by stephenbach** [comment]
Thanks for using Snorkel!

Snorkel uses nondeterministic algorithms like stochastic gradient descent and Gibbs sampling to learn weights, so we expect that changing the order of the input data would produce slight variations in the learned weights. For example, learning weights on a 2,000 example subset of the data in the intro tutorial gives

```
array([0.07816977, 0.07857064, 0.11705587, 0.10540339, 0.07270013,
       0.72678591, 0.0741934 , 0.15670433, 0.0827119 , 0.12854234])
```
(Note that these are the accuracies in log scale so a weight of 0.07 means an estimated accuracy of 1 / (1 + exp(2 * -0.07)) ~ 54%.)

If I switch the last two labeling functions, the estimated weights are

```
array([0.07820237, 0.0765507 , 0.1187966 , 0.10371894, 0.07245944,
       0.72715762, 0.07467522, 0.15667129, 0.1322996 , 0.07666995])
```

Notice how the weights are only slightly different.


So, the big differences that you're seeing suggest that something else is going on. Would you run a few tests to diagnose the issue?

1) Can you confirm that with the labeling functions in the same order, you get the same results every time?
2) Can you confirm that the label matrix is identical up to a permutation of the columns? Sometimes we've seen users write labeling functions that are non-deterministic.
3) Can you let us know how much the weights of the generative model change with different orderings? You can print the arrays I showed above using `gen_model.weights.lf_accuracy` as in the intro. tutorial.

Thanks!

**Comment by tienduccao** [comment]
> 1. Can you confirm that with the labeling functions in the same order, you get the same results every time?

Yes I confirm that.

> 2. Can you confirm that the label matrix is identical up to a permutation of the columns? Sometimes we've seen users write labeling functions that are non-deterministic.

Sorry but I don't get your point here.
As far as I understand, in the label matrix we have each row represents a candidate and each column represents a label that a labelling function gives to a candidate.
So if we do a permutation of the columns, how could the label matrix stay identical?

> 3. Can you let us know how much the weights of the generative model change with different orderings? You can print the arrays I showed above using gen_model.weights.lf_accuracy as in the intro. tutorial.

I got these weights for the scenario 1
`[0.13338967, 0.2869923 , 0.14267186, 0.13781506, 0.13526582, 0.13629561, 0.13137059, 0.14091812, 0.13597307, 0.12896815]`
and these for the scenario 2
`[0.13258404, 0.28681777, 0.14125066, 0.1385395 , 0.13588194, 0.13559558, 0.13145762, 0.13527598, 0.13168602, 0.13358496]`
It seems I had the different weights in the last 3 labelling functions.

**Comment by ajratner** [comment]
Hi @tienduccao to clarify (2): this just means to confirm that the labeling functions each output the same labels regardless of the order they are applied in.  If LFs use random processes, for example, then they could output different labels depending on the order they're applied in

**Comment by tienduccao** [comment]
Hi @ajratner , in my LFs I used dependency parsing from Spacy.
I guess it's not a random process?

**Comment by ajratner** [comment]
Sounds like it shouldn't be but you can just quickly check the LF labels to make sure they stay the same?

**Comment by tienduccao** [comment]
I did a check.
Let `c_1` to `c_10` be the label columns in scenario 1.
Let `d_1` to `d_10` be the label columns in scenario 2.
I have `c_i == d_i` for `i` from 1 to 7.
I also have `c_8 == d_9, c_9 == d_10, c_10 == d_8`.

**Comment by stephenbach** [comment]
Thanks for the helpful info. As you see, there's slight differences in the estimated weights, which we would expect, but I'm surprised that it's leading to such big differences in end model performance.

To further debug, it would be good to check where in the pipeline the big change is happening. The next thing to check is the average difference between the training marginals estimated by one model and the other. If they're small, then for some reason the discriminative model (the TagRNN) is making very different predictions despite only small changes in the training marginals.

A couple of things come to mind that could possibly cause this:
1. The test set is small enough that a few more or less true positives changes the F1 significantly
2. The TagRNN needs stronger regularization

**Comment by tienduccao** [comment]
Thanks for your reply. I'll do the necessary check and let you know.
I only have 50 documents (each document is just a sentence) as test set, do you think it's too small?

**Comment by stephenbach** [comment]
Great! It sounds like (1) is indeed an issue. That number is a little small. We usually try to get a few hundred at least. Can you see how many more or fewer true positives, false positives, and false negatives those changes in score correspond to? If its a very small absolute number, this is probably within expected variation.

**Comment by tienduccao** [comment]
Hi @stephenbach sorry the my late reply.
I increased the amount of test data and the problem is solved now.
Thanks a lot.
